features/distributed/kv-store/cassandra/src/main/java/org/opennms/features/distributed/kvstore/cassandra/CassandraKVStore.java

```diff
@@ -0,0 +1,207 @@
+import org.opennms.features.distributed.cassandra.api.CassandraSessionFactory;
+import org.opennms.features.distributed.kvstore.api.AbstractSerializedKVStore;
+import org.opennms.features.distributed.kvstore.api.SerializedKVStore;
+
+import com.datastax.driver.core.PreparedStatement;
+import com.datastax.driver.core.ResultSet;
+import com.datastax.driver.core.ResultSetFuture;
+import com.datastax.driver.core.Row;
+import com.google.common.util.concurrent.MoreExecutors;
+
+/**
+ * A {@link SerializedKVStore key value store} that is backed by Cassandra.
+ * <p>
+ * This implementation persists values that implement {@link Serializable} by using vanilla Java serialization.
+ * <p>
+ * This implementation persists key values in a specific table that will be created in a keyspace dictated by the
+ * {@link org.opennms.features.distributed.cassandra.api.CassandraSchemaManager schema manager}.
+ * <p>
+ * This implementation does not initiate its own {@link com.datastax.driver.core.Session Cassandra session} and must be
+ * provided with one via a {@link CassandraSessionFactory session factory}.
+ */
+public class CassandraKVStore extends AbstractSerializedKVStore<ByteBuffer, Serializable> {
+    private static final String KEY_COLUMN = "key";
+    private static final String VALUE_COLUMN = "value";
+    private static final String TIMESTAMP_COLUMN = "lastUpdated";
+    private static final String TABLE_NAME = "kvstore";
+
+    private final CassandraSession session;
+    private final PreparedStatement insertStmt;
+    private final PreparedStatement selectStmt;
+    private final PreparedStatement timestampStmt;
+
+    public CassandraKVStore(CassandraSessionFactory sessionFactory,
+                            CassandraSchemaManagerFactory cassandraSchemaManagerFactory) throws IOException {
+        super(CassandraJavaSerializationStrategy.INSTANCE, System::currentTimeMillis);
+        Objects.requireNonNull(sessionFactory);
+        Objects.requireNonNull(cassandraSchemaManagerFactory);
+
+        cassandraSchemaManagerFactory.getSchemaManager()
+                .create(() -> getClass().getResourceAsStream("/cql/kv.cql"));
+        session = sessionFactory.getSession();
+        insertStmt = session.prepare(String.format("INSERT INTO %s (%s, %s, %s) VALUES (?, ?, ?)", TABLE_NAME,
+                KEY_COLUMN,
+                VALUE_COLUMN, TIMESTAMP_COLUMN));
+        selectStmt = session.prepare(String.format("SELECT %s FROM %s where %s = ?", VALUE_COLUMN, TABLE_NAME,
+                KEY_COLUMN));
+        timestampStmt = session.prepare(String.format("SELECT %s FROM %s where %s = ?", TIMESTAMP_COLUMN, TABLE_NAME,
+                KEY_COLUMN));
+    }
+
+    @Override
+    protected void putSerializedValueWithTimestamp(String key, ByteBuffer serializedValue, long timestamp) {
+        // Cassandra will throw a runtime exception here if the execution fails
+        session.execute(insertStmt.bind(key, serializedValue, new Date(timestamp)));
```

Comment:
We should also consider including a TTL and exposing this up through the kvstore API.

i.e. Persist this key, but only keep it for 24 hours unless I write to it again.